File: Sort/MergeSort/merge_sort.py
def merge_sort(arr) :
    if len(arr) > 1:
        r = len(arr)//2
        l = arr[:r]
        m = arr[r:]

        merge_sort(l)
        merge_sort(m)

        i = j = k = 0

        while i < len(l) and j < len(m):
            if l[i] < m[j]:
                arr[k] = l[i]
                i += 1
            else:
                arr[k] = m[j]
                j += 1
            k += 1

        while i < len(l):
            arr[k] = l[i]
            i += 1
            k += 1

        while j < len(m):
            arr[k] = m[j]
            j += 1
            k += 1

given_arr = [6, 5, 4, 3, 2, 1]
merge_sort(given_arr)
print(given_arr)

File: Common Algorithm/Sorting Algorithms/MergeSort/merge_sort.js
const givenArr = [8, 5, 1, 3, 2, 4, 9];
mergeSort(givenArr);
console.log(givenArr);

function mergeSort(arr) {
  if (arr.length <= 1) return;

  const m = Math.floor(arr.length / 2);
  const l = arr.slice(0, m);
  const r = arr.slice(m);

  mergeSort(l);
  mergeSort(r);

  let i = 0;
  let j = 0;
  let k = 0;

  while (i < l.length && j < r.length) {
    if (l[i] < r[j]) {
      arr[k] = l[i];
      i++;
    } else {
      arr[k] = r[j];
      j++;
    }
    k++;
  }

  while (i < l.length) {
    arr[k] = l[i];
    i++;
    k++;
  }

  while (j < r.length) {
    arr[k] = r[j];
    j++;
    k++;
  }
}
